File: 程序员面试金典第6版/面试题 01.06. Compress String LCCI.cpp
class Solution {
public:
    string compressString(string S) {
        string res = "";
        int count=1;
        for(int i=0; i<S.size(); i++) {
            res += S[i];
            while(S[i] == S[i+1]) {
                i++;
                count++;
            }
            res += to_string(count);
            count = 1; // 不要忘记重置count
        }
        if(res.size() >= S.size()) return S;
        return res;
    }
};

File: 程序员面试金典第6版/面试题 01.06. Compress String LCCI.java
class Solution {
    public String compressString(String S) {
        int len = S.length(), count = 1;
        if(len < 2) return S;

        char c = S.charAt(0);
        StringBuilder sb = new StringBuilder();
        int i=1;
        for(;i<len; i++) {
            if(c == S.charAt(i)) {
                count++;
            } else {
                sb.append(c);
                sb.append(count);
                c = S.charAt(i);
                count = 1;
            }
        }
        if(i == len) { // if判断不需要 这步目的是处理最后一个需要压缩的部分
            sb.append(c);
            sb.append(count);
        }
        if(sb.length() < len) {
            return sb.toString();
        } else {
            return S;
        }
    }
}